### ISSUE TITLE
Map-reduce lacks most core properties/functions

### ISSUE THREAD

**Issue by ddossot** [fakemongo/fongo#154]
I'm having issues running map-reduce operations because of the absence of support for `NumberLong` or, for debugging purposes, `print` and `printjson`.

For the record here are the [supported properties/functions in 2.4](https://docs.mongodb.org/v2.6/reference/command/mapReduce/):

![screen shot 2015-10-28 at 12 50 09](https://cloud.githubusercontent.com/assets/119028/10801014/71f6d5ea-7d72-11e5-9e52-6457fd4ed29a.png)

**Comment by twillouer** [comment]
Hi, print, printjson are ok, NumberLong is not good enought. Can you check the v3 branch ?

**Comment by ddossot** [comment]
Just tried and got:

    org.mozilla.javascript.EcmaError: TypeError: Cannot find function toNumber in object 10.

It seems `toNumber` is missing from `NumberLong`. It's one of the method attached to it by Mongo, see: https://github.com/mongodb/mongo/blob/69dec2fe8fed6d32ec4998ea7ec7ab063cb5b788/src/mongo/scripting/mozjs/numberlong.cpp#L47-L50

**Comment by twillouer** [comment]
done

**Comment by ddossot** [comment]
I've done more testing and still experience the same failure. I've tried to use `print` to emit some debugging info but I can't find where does the `print` output go, neither on `sysout` nor `syserr` it seems.

In any case, I believe the issue comes from the fact that some `java.lang.Long` numbers are not converted into `NumberLong` for objects passed via the `scope` mechanism of map/reduce. I have the impression these objects are passed as-is instead of being BSON serialized and deserialized so they end up with the proper Mongo types. I wanted to confirm this with `print`ing out the field types of these `scope`d objects but failed.

**Comment by twillouer** [comment]
Hum... print is normally working and output to stdout.

If it's possible, can you provide your map/reduce scripts ?

**Comment by ddossot** [comment]
Unfortunately, I can't share what I'm working on. I do realize how hard it is to debug issues with the little info I'm providing you :crying_cat_face:

**Comment by twillouer** [comment]
Don't you see any output with "print" ?
Normally, this go to stdout

**Comment by ddossot** [comment]
All I see is the full execution context that's dump when the exception occurs.

**Comment by twillouer** [comment]
Sorry to incist, but 
com.github.fakemongo.FongoMapReduceTest#should_NumberLongtoNumber_works

for example produce an output 

```java
    String reduce = "function(key, values){    var res = 0.0;    values.forEach(function(v){ res += 1.0});    printjson(res); return {\"count\": NumberLong(res).toNumber()};  };";
```

you can see 
```
2
3
```

in the console..

I try to reproduce your problem, but I don't really know where to start...

Can't share your use case by email or gist or something ?

**Comment by ddossot** [comment]
The thing is I do not create `NumberLong` instances myself, they come from Mongo.

For example, if you store longs in an array field, you will receive an array of `NumberLong` instances in map/reduce.  Or if you pass a long via a `scope` global variable, again Mongo will make it available as a `NumberLong` in map/reduce.

So my map/reduce code assumes that these values are `NumberLong` already, and it works fine in Mongo. In Fongo, it seems these values remain `java.lang.Long` instances, which of course don't have `.toNumber`.

Maybe what I'm trying to achieve is not possible, or would be way to complex to implement. Being able to run the same m/r scripts in Mongo and Fongo is no small feat, I do realize the complexity of the task.

**Comment by twillouer** [comment]
again ?

**Comment by ddossot** [comment]
Progress!

I can see that now all longs coming from the query that feeds m/r are wrapped in `NumberLong`s. The only issue left is that long values provided via `scope` are not wrapped and thus calling `toNumber` on them fails.

For testing, I've created a fake `NumberLong` wrapper around a `long` with a `toNumber` method and passed it in `scope` instead of a plain `long` and, success, the whole map/reduce went through without a hitch.

Basically, values (which can be BSON objects) passed via `scope` should be as if they were coming from a DB query, not passed straight as Java objects.

**Comment by twillouer** [comment]
Ho, I forgot the scope :-(

Can you try again ?

**Comment by ddossot** [comment]
Still failing. I've added extra debugging and I see why: `Long` fields of objects passed via scope aren't wrapped with `NumberLong`. This is what I was trying to say when I said:

> values (which can be BSON objects) passed via `scope` should be as if they were coming from a DB query, not passed straight as Java objects.

**Comment by twillouer** [comment]
I have this unit test:

```
    Map scope = new HashMap();
    scope.put("from", 100L);
    scope.put("to", 301L);
    cmd.setScope(scope);
```

```
    String map = "function() {" +
        "if(this.arrayOfStuff) {" +
        "this.arrayOfStuff.forEach(function(stuff) {" +
        "if(stuff.date > from.toNumber() && stuff.date < to.toNumber()) {" +
        "emit({day: stuff.date}, {count:1});" +
        "}" +
        "});" +
        "}" +
        "};";
```
(see com.github.fakemongo.FongoMapReduceTest#should_scope_permit_to_initialize_NumberLong).

What's the difference with yours ?

**Comment by ddossot** [comment]
The difference is mainly:

    scope.put("something", aThing);

where `aThing` is a POJO with `Long` fields.

**Comment by twillouer** [comment]
right. Thanks

**Comment by twillouer** [comment]
again ?

**Comment by ddossot** [comment]
Success! :dancers: Woohoo!

Thank you so much :+1:

**Comment by twillouer** [comment]
![champagne](https://s1.qwant.com/thumbr/?u=http%3A%2F%2Fvenewsbebidas.com.br%2F_imagens%2Fprodutos%2Fchampagne-brut-premier-1500ml_1.jpg)
